### PR TITLE
optimize the map iterator used by the history table

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: go/libraries/doltcore/table/typed/noms/range_reader.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+type InRangeCheck func(tuple types.Tuple) (bool, error)
+
+// ReadRange represents a range of values to be read
+type ReadRange struct {
+	// Start is a Dolt map key which is the starting point (or ending point if Reverse is true)
+	Start types.Tuple
+	// Inclusive says whether the Start key should be included in the range.
+	Inclusive bool
+	// Reverse says if the range should be read in reverse (from high to low) instead of the default (low to high)
+	Reverse bool
```
Comment (zachmu) [inline go/libraries/doltcore/table/typed/noms/range_reader.go]: Call this Ascending / Descending and make an enum type for it

File: go/libraries/doltcore/table/typed/noms/range_reader.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	"io"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/row"
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// InRangeCheck is a call made as the reader reads through values to check that the next value
+// being read is in the range]
+type InRangeCheck func(tuple types.Tuple) (bool, error)
```
Comment (zachmu) [inline go/libraries/doltcore/table/typed/noms/range_reader.go]: What's the advantage to having a range check as opposed to upper and lower bound?

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+		return setForComparisonExp(nbf, col, typedExpr.BinaryExpression, ltOp, setForLtOp)
+	case *expression.LessThanOrEqual:
+		lteOp.NBF = nbf
+		return setForComparisonExp(nbf, col, typedExpr.BinaryExpression, lteOp, setForLteOp)
+	case *expression.In:
+		return setForInExp(nbf, col, typedExpr.BinaryExpression)
+		// case *expression.Subquery:
+	}
+
+	return setalgebra.UniversalSet{}, nil
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: You are correct.  But I don't tell go-mysql-server that I am handling any of these filters.  I only tell go-mysql-server that I'm handling filters on the commits.

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+func getSetForKeyColumn(nbf *types.NomsBinFormat, col schema.Column, filter sql.Expression) (setalgebra.Set, error) {
+	switch typedExpr := filter.(type) {
+	case *expression.Or:
+		return getOrSet(nbf, col, typedExpr.Left, typedExpr.Right)
+	case *expression.And:
+		return getAndSet(nbf, col, typedExpr.Left, typedExpr.Right)
+	case *expression.Equals:
+		return setForComparisonExp(nbf, col, typedExpr.BinaryExpression, eqOp, setForEqOp)
+	case *expression.GreaterThan:
+		gtOp.NBF = nbf
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: fixed

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+	case *expression.In:
+		return setForInExp(nbf, col, typedExpr.BinaryExpression)
+		// case *expression.Subquery:
+	}
+
+	return setalgebra.UniversalSet{}, nil
+}
+
+// getOrSet gets the set of values which satisfies the or condition.  This will be the union of
+// the sets generated by the left and right sides of he or expression.
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: fixed

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+func getOrSet(nbf *types.NomsBinFormat, col schema.Column, left, right sql.Expression) (setalgebra.Set, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: changed

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return leftSet.Union(rightSet)
+}
+
+// getAndSet gets the set of values which satisfies the and condition.  This will be the intersection of
+// the sets generated by the left and right sides of he or expression.
+func getAndSet(nbf *types.NomsBinFormat, col schema.Column, left, right sql.Expression) (setalgebra.Set, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: changed

File: go/libraries/doltcore/sqle/expression_limited_map_reader.go [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+	return leftSet.Intersect(rightSet)
+}
+
+// setForComparisonExp returns the set of values which satisfies the comparison expression for the specified column.
+func setForComparisonExp(
+	nbf *types.NomsBinFormat,
+	col schema.Column,
+	be expression.BinaryExpression,
+	op expreval.CompareOp,
+	createSet func(nbf *types.NomsBinFormat, val types.Value) (setalgebra.Set, error)) (setalgebra.Set, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/sqle/expression_limited_map_reader.go]: created a type for the function which makes it more readable.

File: go/libraries/doltcore/table/typed/noms/range_reader.go (hunk 2 above)
Comment (bheni) [reply go/libraries/doltcore/table/typed/noms/range_reader.go]: There is no advantage yet, but when dealing with multi column keys it provides more flexibility, and it makes it trivial to write a default check function which just compares against an end value if that's the behavior you want.

**Pull request merged**